File: src/model_detector.py
"""
Автоопределение доступных LLM моделей и провайдеров.
"""
from __future__ import annotations

import os
from dataclasses import dataclass

try:
    import requests
    HAS_REQUESTS = True
except ImportError:
    HAS_REQUESTS = False


@dataclass
class AvailableModel:
    """Доступная модель."""
    provider: str
    model: str
    name: str
    size: str | None = None
    speed: str = "medium"  # fast, medium, slow


def detect_ollama_models() -> list[AvailableModel]:
    """Определить доступные модели Ollama."""
    if not HAS_REQUESTS:
        return []

    try:
        response = requests.get("http://localhost:11434/api/tags", timeout=2)
        if response.status_code == 200:
            data = response.json()
            models = []
            for model_info in data.get("models", []):
                name = model_info.get("name", "")
                size = model_info.get("details", {}).get("parameter_size", "")

                # Определяем скорость по размеру
                speed = "medium"
                if "1.5b" in name.lower() or "1b" in name.lower():
                    speed = "fast"
                elif "7b" in name.lower():
                    speed = "medium"
                elif "13b" in name.lower() or "70b" in name.lower():
                    speed = "slow"

                models.append(AvailableModel(
                    provider="ollama",
                    model=name,
                    name=f"Ollama: {name}",
                    size=size,
                    speed=speed
                ))
            return models
    except Exception:
        pass

    return []


def detect_anthropic_available() -> list[AvailableModel]:
    """Проверить доступность Anthropic API."""
    if os.getenv("ANTHROPIC_API_KEY"):
        return [
            AvailableModel(
                provider="anthropic",
                model="claude-3-5-sonnet-20241022",
                name="Anthropic: Claude 3.5 Sonnet",
                speed="fast"
            ),
            AvailableModel(
                provider="anthropic",
                model="claude-3-opus-20240229",
                name="Anthropic: Claude 3 Opus",
                speed="medium"
            ),
        ]
    return []


def detect_openai_available() -> list[AvailableModel]:
    """Проверить доступность OpenAI API."""
    if os.getenv("OPENAI_API_KEY"):
        return [
            AvailableModel(
                provider="openai",
                model="gpt-4o",
                name="OpenAI: GPT-4o",
                speed="fast"
            ),
            AvailableModel(
                provider="openai",
                model="gpt-4o-mini",
                name="OpenAI: GPT-4o Mini",
                speed="fast"
            ),
            AvailableModel(
                provider="openai",
                model="gpt-4-turbo",
                name="OpenAI: GPT-4 Turbo",
                speed="medium"
            ),
        ]
    return []


def detect_openrouter_available() -> list[AvailableModel]:
    """Проверить доступность OpenRouter API."""
    if os.getenv("OPENROUTER_API_KEY"):
        # Получаем модель из переменной окружения или используем дефолтную
        model = os.getenv("OPENROUTER_MODEL", "qwen/qwen-2.5-coder-32b-instruct")

        # Определяем скорость по названию модели
        speed = "fast"
        if "free" in model.lower():
            speed = "fast"
        elif "32b" in model.lower() or "70b" in model.lower():
            speed = "medium"
        elif "405b" in model.lower():
            speed = "slow"

        return [
            AvailableModel(
                provider="openrouter",
                model=model,
                name=f"OpenRouter: {model}",
                speed=speed
            ),
        ]
    return []


def detect_koboldcpp_available() -> list[AvailableModel]:
    """Проверить доступность KoboldCPP сервера."""
    if not HAS_REQUESTS:
        return []

    # Используем умную систему автопоиска
    try:
        from .koboldcpp_discovery import find_koboldcpp_server
        koboldcpp_url = find_koboldcpp_server(timeout=2.0)

        if koboldcpp_url:
            try:
                response = requests.get(f"{koboldcpp_url}/api/v1/model", timeout=2)
                if response.status_code == 200:
                    data = response.json()
                    model_name = data.get("result", "local-model")

                    return [
                        AvailableModel(
                            provider="koboldcpp",
                            model=model_name,
                            name=f"KoboldCPP: {model_name}",
                            speed="fast"
                        )
                    ]
            except Exception:
                pass

    except ImportError:
        # Fallback если модуль discovery не доступен
        pass

    # Fallback: проверяем известные адреса
    possible_urls = [
        os.getenv("KOBOLDCPP_URL", "http://localhost:5001"),
        "http://localhost:5001",
        "http://192.168.1.136:5001",
        "http://127.0.0.1:5001",
    ]

    for koboldcpp_url in possible_urls:
        try:
            response = requests.get(f"{koboldcpp_url}/api/v1/model", timeout=2)
            if response.status_code == 200:
                data = response.json()
                model_name = data.get("result", "local-model")

                return [
                    AvailableModel(
                        provider="koboldcpp",
                        model=model_name,
                        name=f"KoboldCPP: {model_name}",
                        speed="fast"
                    )
                ]
        except Exception:
            continue

    return []



def detect_all_models() -> list[AvailableModel]:
    """Определить все доступные модели с приоритетом на локальные."""
    models = []
    
    # 1. KoboldCPP (локальный сервер - ВЫСШИЙ ПРИОРИТЕТ)
    kobold_models = detect_koboldcpp_available()
    if kobold_models:
        models.extend(kobold_models)
        print(f"✅ Найдено KoboldCPP моделей: {len(kobold_models)}")
    
    # 2. Ollama (локальные модели)
    ollama_models = detect_ollama_models()
    if ollama_models:
        models.extend(ollama_models)
        print(f"✅ Найдено Ollama моделей: {len(ollama_models)}")
    
    # 3. OpenRouter (только если нет локальных)
    if not models:
        openrouter_models = detect_openrouter_available()
        if openrouter_models:
            models.extend(openrouter_models)
            print(f"✅ Найдено OpenRouter моделей: {len(openrouter_models)}")
    
    # 4. API провайдеры (только если нет локальных)
    if not models:
        models.extend(detect_anthropic_available())
        models.extend(detect_openai_available())
    
    return models

def get_recommended_model() -> AvailableModel | None:
    """Получить рекомендуемую модель (KoboldCPP > Ollama > OpenRouter > API)."""
    models = detect_all_models()
    
    if not models:
        return None
    
    # Приоритет: KoboldCPP > Ollama > OpenRouter > Anthropic > OpenAI
    for model in models:
        if model.provider == "koboldcpp":
            return model
    
    for model in models:
        if model.provider == "ollama":
            return model
    
    for model in models:
        if model.provider == "openrouter":
            return model
    
    for model in models:
        if model.provider == "anthropic":
            return model
    
    return models[0]
    """Получить самую быструю доступную модель."""
    models = detect_all_models()

    if not models:
        return None

    # Сортируем: сначала fast, потом medium, потом slow
    speed_order = {"fast": 0, "medium": 1, "slow": 2}
    models.sort(key=lambda m: speed_order.get(m.speed, 3))

    return models[0]


def get_recommended_model() -> AvailableModel | None:
    """Получить рекомендуемую модель (баланс скорости и качества)."""
    models = detect_all_models()

    if not models:
        return None

    # Приоритет: OpenRouter > Anthropic > OpenAI > Ollama medium > Ollama fast
    for model in models:
        if model.provider == "openrouter":
            return model

    for model in models:
        if model.provider == "anthropic":
            return model

    for model in models:
        if model.provider == "openai":
            return model

    for model in models:
        if model.provider == "ollama" and model.speed == "medium":
            return model

    return models[0]


def print_available_models():
    """Вывести список доступных моделей."""
    models = detect_all_models()

    if not models:
        print("❌ Нет доступных моделей")
        print("\nУстановите:")
        print("  - Ollama: https://ollama.com")
        print("  - OpenRouter API: export OPENROUTER_API_KEY='sk-or-v1-...'")
        print("  - Anthropic API: export ANTHROPIC_API_KEY='sk-ant-...'")
        print("  - OpenAI API: export OPENAI_API_KEY='sk-...'")
        return

    print(f"✅ Найдено моделей: {len(models)}\n")

    # Группируем по провайдерам
    by_provider = {}
    for model in models:
        if model.provider not in by_provider:
            by_provider[model.provider] = []
        by_provider[model.provider].append(model)

    for provider, provider_models in by_provider.items():
        print(f"📦 {provider.upper()}:")
        for model in provider_models:
            speed_emoji = {"fast": "⚡", "medium": "🔄", "slow": "🐌"}.get(model.speed, "❓")
            size_info = f" ({model.size})" if model.size else ""
            print(f"   {speed_emoji} {model.model}{size_info}")
        print()

    # Рекомендации
    fastest = get_fastest_model()
    recommended = get_recommended_model()

    if fastest:
        print(f"⚡ Самая быстрая: {fastest.name}")
    if recommended and recommended != fastest:
        print(f"⭐ Рекомендуемая: {recommended.name}")


if __name__ == "__main__":
    print_available_models()

File: src/model_detector.rs
//! Автоопределение доступных LLM провайдеров/моделей. Порт
//! `model_detector.py` (итерация 6). У оригинала было два конфликтующих
//! определения `get_recommended_model()` (Python просто использует
//! последнее — первое было мёртвым кодом с недостижимым `return`
//! посередине); здесь только финальная, реально используемая логика.
//! `koboldcpp_discovery.py` (умный автопоиск по подсети) не портирован —
//! используется прямой fallback-список адресов, как и в самом
//! `model_detector.py` при `ImportError`.

use serde::Serialize;
use std::env;
use std::time::Duration;

#[derive(Debug, Clone, Serialize)]
pub struct AvailableModel {
    pub provider: String,
    pub model: String,
    pub name: String,
    pub size: Option<String>,
    pub speed: String, // fast | medium | slow
}

fn client() -> reqwest::Client {
    reqwest::Client::builder()
        .timeout(Duration::from_secs(2))
        .build()
        .unwrap_or_default()
}

async fn detect_ollama_models() -> Vec<AvailableModel> {
    let c = client();
    let resp = match c.get("http://localhost:11434/api/tags").send().await {
        Ok(r) if r.status().is_success() => r,
        _ => return vec![],
    };
    let data: serde_json::Value = match resp.json().await {
        Ok(v) => v,
        Err(_) => return vec![],
    };
    let mut models = vec![];
    if let Some(arr) = data.get("models").and_then(|v| v.as_array()) {
        for m in arr {
            let name = m.get("name").and_then(|v| v.as_str()).unwrap_or("").to_string();
            let size = m
                .get("details")
                .and_then(|d| d.get("parameter_size"))
                .and_then(|v| v.as_str())
                .map(|s| s.to_string());
            let lower = name.to_lowercase();
            let speed = if lower.contains("1.5b") || lower.contains("1b") {
                "fast"
            } else if lower.contains("13b") || lower.contains("70b") {
                "slow"
            } else {
                "medium"
            };
            models.push(AvailableModel {
                provider: "ollama".to_string(),
                model: name.clone(),
                name: format!("Ollama: {}", name),
                size,
                speed: speed.to_string(),
            });
        }
    }
    models
}

fn detect_anthropic_available() -> Vec<AvailableModel> {
    if env::var("ANTHROPIC_API_KEY").is_ok() {
        vec![
            AvailableModel {
                provider: "anthropic".to_string(),
                model: "claude-3-5-sonnet-20241022".to_string(),
                name: "Anthropic: Claude 3.5 Sonnet".to_string(),
                size: None,
                speed: "fast".to_string(),
            },
            AvailableModel {
                provider: "anthropic".to_string(),
                model: "claude-3-opus-20240229".to_string(),
                name: "Anthropic: Claude 3 Opus".to_string(),
                size: None,
                speed: "medium".to_string(),
            },
        ]
    } else {
        vec![]
    }
}

fn detect_openai_available() -> Vec<AvailableModel> {
    if env::var("OPENAI_API_KEY").is_ok() {
        vec![
            AvailableModel {
                provider: "openai".to_string(),
                model: "gpt-4o".to_string(),
                name: "OpenAI: GPT-4o".to_string(),
                size: None,
                speed: "fast".to_string(),
            },
            AvailableModel {
                provider: "openai".to_string(),
                model: "gpt-4o-mini".to_string(),
                name: "OpenAI: GPT-4o Mini".to_string(),
                size: None,
                speed: "fast".to_string(),
            },
            AvailableModel {
                provider: "openai".to_string(),
                model: "gpt-4-turbo".to_string(),
                name: "OpenAI: GPT-4 Turbo".to_string(),
                size: None,
                speed: "medium".to_string(),
            },
        ]
    } else {
        vec![]
    }
}

fn detect_openrouter_available() -> Vec<AvailableModel> {
    match env::var("OPENROUTER_API_KEY") {
        Ok(_) => {
            let model = env::var("OPENROUTER_MODEL")
                .unwrap_or_else(|_| "qwen/qwen-2.5-coder-32b-instruct".to_string());
            let lower = model.to_lowercase();
            let speed = if lower.contains("405b") {
                "slow"
            } else if lower.contains("32b") || lower.contains("70b") {
                "medium"
            } else {
                "fast"
            };
            vec![AvailableModel {
                provider: "openrouter".to_string(),
                model: model.clone(),
                name: format!("OpenRouter: {}", model),
                size: None,
                speed: speed.to_string(),
            }]
        }
        Err(_) => vec![],
    }
}

async fn detect_koboldcpp_available() -> Vec<AvailableModel> {
    let c = client();
    let mut candidates = vec![env::var("KOBOLDCPP_URL").unwrap_or_else(|_| "http://localhost:5001".to_string())];
    for u in ["http://localhost:5001", "http://192.168.1.136:5001", "http://127.0.0.1:5001"] {
        if !candidates.contains(&u.to_string()) {
            candidates.push(u.to_string());
        }
    }

    for url in candidates {
        let resp = match c.get(format!("{}/api/v1/model", url)).send().await {
            Ok(r) if r.status().is_success() => r,
            _ => continue,
        };
        if let Ok(data) = resp.json::<serde_json::Value>().await {
            let model_name = data.get("result").and_then(|v| v.as_str()).unwrap_or("local-model").to_string();
            return vec![AvailableModel {
                provider: "koboldcpp".to_string(),
                model: model_name.clone(),
                name: format!("KoboldCPP: {}", model_name),
                size: None,
                speed: "fast".to_string(),
            }];
        }
    }
    vec![]
}

/// Определить все доступные модели с приоритетом на локальные
/// (KoboldCPP/Ollama > облачные API, если локальных не найдено).
pub async fn detect_all_models() -> Vec<AvailableModel> {
    let mut models = vec![];

    let kobold = detect_koboldcpp_available().await;
    if !kobold.is_empty() {
        models.extend(kobold);
    }

    let ollama = detect_ollama_models().await;
    if !ollama.is_empty() {
        models.extend(ollama);
    }

    if models.is_empty() {
        let openrouter = detect_openrouter_available();
        if !openrouter.is_empty() {
            models.extend(openrouter);
        }
    }

    if models.is_empty() {
        models.extend(detect_anthropic_available());
        models.extend(detect_openai_available());
    }

    models
}

/// Самая быстрая доступная модель (fast > medium > slow).
pub fn fastest(models: &[AvailableModel]) -> Option<&AvailableModel> {
    fn rank(speed: &str) -> u8 {
        match speed {
            "fast" => 0,
            "medium" => 1,
            _ => 2,
        }
    }
    models.iter().min_by_key(|m| rank(&m.speed))
}

/// Рекомендуемая модель — баланс скорости и качества. Приоритет (как в
/// финальной, реально используемой версии `model_detector.py`):
/// OpenRouter > Anthropic > OpenAI > Ollama(medium) > первая из списка.
pub fn recommended(models: &[AvailableModel]) -> Option<&AvailableModel> {
    if models.is_empty() {
        return None;
    }
    for provider in ["openrouter", "anthropic", "openai"] {
        if let Some(m) = models.iter().find(|m| m.provider == provider) {
            return Some(m);
        }
    }
    if let Some(m) = models.iter().find(|m| m.provider == "ollama" && m.speed == "medium") {
        return Some(m);
    }
    models.first()
}

/// Отформатированный вывод (аналог `print_available_models()`), но
/// возвращает строку вместо печати — вызывающий код сам решает, куда её
/// вывести (REPL, лог, инструмент).
pub async fn describe_available_models() -> String {
    let models = detect_all_models().await;
    if models.is_empty() {
        return "❌ Нет доступных моделей\n\nУстановите:\n  - Ollama: https://ollama.com\n  \
                - OpenRouter API: export OPENROUTER_API_KEY='sk-or-v1-...'\n  \
                - Anthropic API: export ANTHROPIC_API_KEY='sk-ant-...'\n  \
                - OpenAI API: export OPENAI_API_KEY='sk-...'"
            .to_string();
    }

    let mut out = format!("✅ Найдено моделей: {}\n\n", models.len());
    let mut by_provider: std::collections::BTreeMap<&str, Vec<&AvailableModel>> = std::collections::BTreeMap::new();
    for m in &models {
        by_provider.entry(m.provider.as_str()).or_default().push(m);
    }
    for (provider, ms) in by_provider {
        out.push_str(&format!("📦 {}:\n", provider.to_uppercase()));
        for m in ms {
            let icon = match m.speed.as_str() {
                "fast" => "⚡",
                "medium" => "🔄",
                "slow" => "🐌",
                _ => "❓",
            };
            let size_info = m.size.as_ref().map(|s| format!(" ({})", s)).unwrap_or_default();
            out.push_str(&format!("   {} {}{}\n", icon, m.model, size_info));
        }
        out.push('\n');
    }

    if let Some(f) = fastest(&models) {
        out.push_str(&format!("⚡ Самая быстрая: {}\n", f.name));
    }
    if let Some(r) = recommended(&models) {
        if fastest(&models).map(|f| &f.name) != Some(&r.name) {
            out.push_str(&format!("⭐ Рекомендуемая: {}\n", r.name));
        }
    }
    out
}
